README.md

```diff
@@ -291,6 +289,331 @@ Available core tools:
+
+- **Over-exposure**: Servers receive environment variables they don't need
+- **Credential leakage**: Sensitive tokens intended for other services may be exposed
+- **Attack surface**: Each server has access to more credentials than necessary
+
+MCP Funnel's secret provider architecture follows the principle of least privilege, ensuring each server only receives the environment variables it actually needs.
+
+### Architecture Overview
+
+The secret management system is built around a modular provider architecture:
+
+```
+┌─────────────────────┐
+│  Configuration      │
+│  secretProviders: [ │
+│    { type: "dotenv" }│
+│    { type: "process"}│
+│    { type: "inline"} │
+│  ]                  │
+└──────────┬──────────┘
+           │
+    ┌──────▼──────┐
+    │ SecretManager│ ← Orchestrates providers, handles precedence
+    └──────┬──────┘
+           │
+   ┌───────┼───────┬─────────┐
+   │       │       │         │
+┌──▼────┐ ┌▼─────┐ ┌▼──────┐ │
+│DotEnv │ │Process│ │Inline │ │ ← Each provider resolves from its source
+│Provider│ │Provider│ │Provider│ │
+└───────┘ └──────┘ └───────┘ │
+   │       │       │         │
+   ▼       ▼       ▼         ▼
+┌─────────────────────────────┐
+│    Merged Environment       │ ← Later providers override earlier ones
+│  { API_KEY: "...",         │
+│    NODE_ENV: "production" } │
+└─────────────────────────────┘
+```
+
+### Available Provider Types
+
+MCP Funnel supports three types of secret providers:
+
+#### 1. DotEnv Provider (`type: "dotenv"`)
+
+Loads secrets from `.env` files on the filesystem. This is the most common approach for managing API tokens and other sensitive configuration.
+
+**Configuration:**
+
+- `path`: Path to the .env file (relative to config file or absolute)
+- `encoding`: File encoding (default: 'utf-8')
+
+**Use cases:**
+
+- GitHub tokens, database URLs, API keys
+- Environment-specific configuration (`.env.development`, `.env.production`)
+- Keeping secrets out of version control
+
+#### 2. Process Environment Provider (`type: "process"`)
+
+Filters and forwards environment variables from the current process. Provides fine-grained control over which variables are passed through.
+
+**Configuration:**
+
+- `prefix`: Include only variables starting with this prefix (prefix is stripped)
+- `allowlist`: Explicit list of variable names to include
+- `blocklist`: Explicit list of variable names to exclude
+
+**Use cases:**
+
+- CI/CD environments where secrets are injected as environment variables
+- Filtering system variables vs application variables
+- Namespace-based organization (e.g., `MCP_API_KEY`)
+
+#### 3. Inline Provider (`type: "inline"`)
+
+Provides static key-value pairs directly in the configuration. **Use with caution** as values are stored in plain text.
+
+**Configuration:**
+
+- `values`: Object with key-value pairs of secrets
+
+**Use cases:**
+
```

Comment:
This example config won’t actually validate: the inline provider schema requires `config.values`, but the snippet shows `"config": { "API_KEY": "static-value" }`. Users who copy this will get a schema error (and the provider won’t load). Please update the docs to match the runtime shape, e.g. `"config": { "values": { "API_KEY": "static-value" } }`.